From 44ef0c06f61d77f5cd666efd858c8b36283016bc Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 26 Aug 2014 14:11:26 -0400
Subject: [ethernet] igb: Avoid uninitialized advertised variable in eee_set_cur

Message-id: <1409062359-13063-9-git-send-email-sassmann@redhat.com>
Patchwork-id: 89127
O-Subject: [RHEL7.1 PATCH 08/81] igb: Avoid uninitialized advertised variable in eee_set_cur
Bugzilla: 1091121
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Andi Kleen <ak@linux.intel.com>

eee_get_cur assumes that the output data is already zeroed. It can
read-modify-write the advertised field:

              if (ipcnfg & E1000_IPCNFG_EEE_100M_AN)
2594   edata->advertised |= ADVERTISED_100baseT_Full;

This is ok for the normal ethtool eee_get call, which always
zeroes the input data before.

But eee_set_cur also calls eee_get_cur and it did not zero the input
field. Later on it then compares agsinst the field, which can contain partial
stack garbage.

Zero the input field in eee_set_cur() too.

Cc: jeffrey.t.kirsher@intel.com
Cc: netdev@vger.kernel.org
Signed-off-by: Andi Kleen <ak@linux.intel.com>
Acked-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 58e4e1f6cacddb7823c44bcfb272174553f6c645)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/igb/igb_ethtool.c b/drivers/net/ethernet/intel/igb/igb_ethtool.c
index ebdac02..0ae3177 100644
--- a/drivers/net/ethernet/intel/igb/igb_ethtool.c
+++ b/drivers/net/ethernet/intel/igb/igb_ethtool.c
@@ -2657,6 +2657,8 @@ static int igb_set_eee(struct net_device *netdev,
      (hw->phy.media_type != e1000_media_type_copper))
   return -EOPNOTSUPP;
 
+ memset(&eee_curr, 0, sizeof(struct ethtool_eee));
+
  ret_val = igb_get_eee(netdev, &eee_curr);
  if (ret_val)
   return ret_val;
-- 
1.7.1